src/subcommand/torrent/announce.rs

```diff
@@ -0,0 +1,135 @@
+use crate::common::*;
+
+const URI_HELP: &str = "Announce an infohash and list the response (a compact peer list).";
+
+const INPUT_HELP: &str =
+  "Generate a compact peer list from a metainfo at `INPUT`. If `INPUT` is `-`, read \
+                          metainfo from standard input.";
+
+const INPUT_FLAG: &str = "input-flag";
+
+const INPUT_POSITIONAL: &str = "<INPUT>";
+
+#[derive(StructOpt)]
+#[structopt(
+  help_message(consts::HELP_MESSAGE),
+  version_message(consts::VERSION_MESSAGE),
+  about(URI_HELP)
+)]
+pub(crate) struct Announce {
+  #[structopt(
+    name = INPUT_FLAG,
+    long = "input",
+    short = "i",
+    value_name = "INPUT",
+    empty_values(false),
+    parse(try_from_os_str = InputTarget::try_from_os_str),
+    help = INPUT_HELP,
+  )]
+  input_flag: Option<InputTarget>,
+  #[structopt(
+    name = INPUT_POSITIONAL,
+    value_name = "INPUT",
+    empty_values(false),
+    parse(try_from_os_str = InputTarget::try_from_os_str),
+    required_unless = INPUT_FLAG,
+    conflicts_with = INPUT_FLAG,
+    help = INPUT_HELP,
+  )]
+  input_positional: Option<InputTarget>,
+}
+
+impl Announce {
+  pub(crate) fn run(self, env: &mut Env, options: &Options) -> Result<(), Error> {
+    let target = xor_args(
+      "input_flag",
+      &self.input_flag,
+      "input_positional",
+      &self.input_positional,
+    )?;
+    let input = env.read(target)?;
+
+    let infohash = Infohash::from_input(&input)?;
+    let metainfo = Metainfo::from_input(&input)?;
+
+    let mut rng = rand::thread_rng();
```

Comment:
I think that `rng` should be part of env. This can be used to make RNG calls deterministic during testing, which seems useful. However, it's definitely something that should be done in a different PR, so don't worry about it. I opened #470 to track this.